trace/tracestate.go

```diff
@@ -0,0 +1,100 @@
+// Copyright 2018, OpenCensus Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package trace
+
+import (
+	"fmt"
+	"regexp"
+)
+
+const (
+	KEY_WITHOUT_VENDOR_FORMAT = `[a-z][_0-9a-z\-\*\/]{0,255}`
+	KEY_WITH_VENDOR_FORMAT    = `[a-z][_0-9a-z\-\*\/]{0,240}@[a-z][_0-9a-z\-\*\/]{0,13}`
+	KEY_FORMAT                = `(` + KEY_WITHOUT_VENDOR_FORMAT + `)|(` + KEY_WITH_VENDOR_FORMAT + `)`
+	VALUE_FORMAT              = `[\x20-\x2b\x2d-\x3c\x3e-\x7e]{0,255}[\x21-\x2b\x2d-\x3c\x3e-\x7e]`
+)
+
+var KEY_VALIDATION_RE = regexp.MustCompile(`^(` + KEY_FORMAT + `)$`)
+var VALUE_VALIDATION_RE = regexp.MustCompile(`^(` + VALUE_FORMAT + `)$`)
+
+type TracestateEntry struct {
+	key   string
+	value string
+}
+
+func (ts *TracestateEntry) IsValid() bool {
+	return KEY_VALIDATION_RE.MatchString(ts.key) &&
+		VALUE_VALIDATION_RE.MatchString(ts.value)
+}
+
+// TraceState is a tracing-system specific context in a list of key-value pairs.
+// TraceState allows different vendors propagate additional information and
+// inter-operate with their legacy Id formats.
+type Tracestate struct {
```

Comment:
Why is this not just a map?

Reply:
because order of key-value pairs is important, see: 
https://w3c.github.io/distributed-tracing/trace_context/HTTP_HEADER_FORMAT_RATIONALE.html#ordering-of-keys-in-tracestate

Reply:
@ramonza - @basvanbeek is right, the W3C spec requires tracestate to following a special ordering rule - last modified key-value-pair should appear at the very beginning.

Reply:
@reyang I think with @ramonza questioning the choice I think it makes sense to add a comment in the code so it is clear to everybody without needing to deep dive into the specs.